pkg/server/v2/server.go

```diff
@@ -136,381 +118,20 @@ func (c CallbackFuncs) OnFetchResponse(req *discovery.DiscoveryRequest, resp *di
 
 // NewServer creates handlers from a config watcher and callbacks.
 func NewServer(ctx context.Context, config cache.Cache, callbacks Callbacks) Server {
-	return &server{cache: config, callbacks: callbacks, ctx: ctx}
+	return NewServerAdvanced(rest.NewServer(config, callbacks), sotw.NewServer(ctx, config, callbacks))
 }
 
-type server struct {
-	cache     cache.Cache
-	callbacks Callbacks
-
-	// streamCount for counting bi-di streams
-	streamCount int64
-	ctx         context.Context
-}
-
-// Generic RPC stream.
-type Stream interface {
-	grpc.ServerStream
-
-	Send(*discovery.DiscoveryResponse) error
-	Recv() (*discovery.DiscoveryRequest, error)
-}
-
-// watches for all xDS resource types
-type watches struct {
-	endpoints chan cache.Response
-	clusters  chan cache.Response
-	routes    chan cache.Response
-	listeners chan cache.Response
-	secrets   chan cache.Response
-	runtimes  chan cache.Response
-
-	endpointCancel func()
-	clusterCancel  func()
-	routeCancel    func()
-	listenerCancel func()
-	secretCancel   func()
-	runtimeCancel  func()
-
-	endpointNonce string
-	clusterNonce  string
-	routeNonce    string
-	listenerNonce string
-	secretNonce   string
-	runtimeNonce  string
-
-	// Opaque resources share a muxed channel. Nonces and watch cancellations are indexed by type URL.
-	responses     chan cache.Response
-	cancellations map[string]func()
-	nonces        map[string]string
-	terminations  map[string]chan struct{}
-}
-
-// Initialize all watches
-func (values *watches) Init() {
-	// muxed channel needs a buffer to release go-routines populating it
-	values.responses = make(chan cache.Response, 5)
-	values.cancellations = make(map[string]func())
-	values.nonces = make(map[string]string)
-	values.terminations = make(map[string]chan struct{})
-}
-
-// Token response value used to signal a watch failure in muxed watches.
-var errorResponse = &cache.RawResponse{}
-
-// Cancel all watches
-func (values *watches) Cancel() {
-	if values.endpointCancel != nil {
-		values.endpointCancel()
-	}
-	if values.clusterCancel != nil {
-		values.clusterCancel()
-	}
-	if values.routeCancel != nil {
-		values.routeCancel()
-	}
-	if values.listenerCancel != nil {
-		values.listenerCancel()
-	}
-	if values.secretCancel != nil {
-		values.secretCancel()
-	}
-	if values.runtimeCancel != nil {
-		values.runtimeCancel()
-	}
-	for _, cancel := range values.cancellations {
-		if cancel != nil {
-			cancel()
-		}
-	}
-	for _, terminate := range values.terminations {
-		close(terminate)
-	}
+func NewServerAdvanced(restServer rest.Server, sotwServer sotw.Server) Server {
+	return &server{rest: restServer, sotw: sotwServer}
```

Comment:
Why a wrapper constructor here? I'd prefer something like a functional opts pattern here if we're going to keep the separate callback functionality

```go
type ServerOpts func(*Server)

func WithSotwCallbacks(callbacks sotw.Callbacks) ServerOpts {
	return func(s *Server) {
		s.SotwCallbacks = callbacks
	}
}

func WithRestCallbacks(callbacks rest.Callbacks) ServerOpts {
	return func(s *Server) {
		s.RestCallbacks = callbacks
	}
}

func NewServer(ctx, config, WithSotwCallbacks(), WithRestCallbacks()) {
... do our opts processing here
}
```

Just in case you haven't seen the pattern before here's a good example of how to do so:
https://www.sohamkamani.com/golang/options-pattern/

Reply:
It doesn't make any sense. Server by itself does not need any ctx, config, cause server itself consists of 2 extracted "subservers". There is an existing API of NewServer that we cannot break and it doesn't make any sense to extend it.

My options were - either this new method or expose `server` type and it's fields externally

Reply:
I concur with @fxposter that maintaining the signature of `NewServer` means fewer breakages so should be preferred.

That said, a functional opts pattern, if desired, could be used in the `NewServerAdvanced` implementation but I think it would make most sense if we expect that servers would be instantiated without one or more of the specific server implementations (i.e., REST, SotW or Delta).

Reply:
@fxposter I see what you're saying. Unless we choose to break the original NewServer function but I don't know if the maintainers want to do that. If we want I could make this change in my incremental PR since it's already breaking API changes.

Reply:
Agreed @joshua-rutherford my reasoning behind the functional opts was to have less code needed to create the needed servers since some users might not want all the server types.